Quaternion/Quaternion.py

```diff
@@ -247,11 +262,16 @@ def _set_q(self, q):
 
         :param q: list or array of normalized quaternion elements
         """
-        q = np.atleast_2d(np.array(q))
+        q = np.atleast_2d(np.array(q))  # asarray() to use a ref???
         if np.any((np.sum(q ** 2, axis=-1, keepdims=True) - 1.0) > 1e-6):
             raise ValueError(
                 'Quaternions must be normalized so sum(q**2) == 1; use Quaternion.normalize')
-        self._q = q
+
+        if self._q is None:
+            self._q = q
+        else:
+            # Set in-place if already set.
+            self._q[...] = q
```

Comment:
This is correct only if the shape of q is the same as the shape of self._q. For example, the following:
```
q = Quat(equatorial=[ 30.000008,  39.999999,  49.999995])
q2 = Quat(equatorial=[[ 30.000008,  39.999999,  49.999995], [ 30.000016,  39.999999,  49.99999 ]])
q.q = q2.q
print(q)
```
gives me the error:
```
ValueError: cannot reshape array of size 8 into shape (4,)
```

Reply:
There was a reason I originally did that (as in something didn't work right), but now I can't reproduce any problems if I fall back to the original logic.  So 1ebc8cf reverts this, good catch. But each of the 3 setters also needs a line to set the shape correctly since it might change, and so I added that.

In addition (with apologies for not making an intermediate commit with just that change), I also took the opportunity to use`@property` as a decorator everywhere for the get/set methods. I think this is generally the preferred idiom. Because the original code had the set before get everywhere, this means the diffs are messy.

Reply:
sorry if I am adding noise, but looking at this commit now, I see that in the setters you set the shape of q before `atleast_2d`, and for equatorial and transform you do it after:
```
        q = np.array(q)
        shape = q.shape[:-1]
        q = np.atleast_2d(q)
```
equatorial:
```
        equatorial = np.array(equatorial)
        self._equatorial = np.atleast_2d(equatorial)
        ...
        self._shape = equatorial.shape[:-1]
```
transform:
```
        self._T = np.atleast_3d(transform)
        ...
        self._shape = transform.shape[:-2]
```
Right now I would have to think a minute more, but it seems to me that it should always be after `atleast_*` to guarantee the shape is standard. If all tests passed, I am guessing there is an untested case

Reply:
In all 3 cases the shape corresponds to the shape of the original input. I put the actual setting of `self._shape` at the end just to minimize object corruption in case of an exception along the way.

The point is that if the user provides `equatorial=[1,2,3]` then we want `shape = ()`. If you get the shape after the `atleast_2d` then the shape is `(1,)` which is not correct.

I think the code will be nicer if I do set a `shape` variable consistently at the beginning for all 3 cases, so I'll do that.

Reply:
Done in 79ff7aa.

Reply:
ah, right.